src/qibo/measurements.py

```diff
@@ -96,6 +98,11 @@ def __repr__(self):
         nshots = self.nshots
         return f"MeasurementResult(qubits={qubits}, nshots={nshots})"
 
+    @property
+    def raw(self) -> dict:
+        samples = self._samples.tolist() if self.has_samples() else self._samples
+        return {"samples": samples}
```

Comment:
In this way, you're discarding all other attributes, including public ones (`.measurement_gate`, `.backend`, `.nshots`, `.circuit`).

Isn't it a bit poor for the serialization?

I understand that all the complex objects' references are hard to serialize (and, arguably, they should not be there in the first place...), but at least `.nshots` should be a simple object (not sure if it's worth to serialize the backend through its name...).

Reply:
Well, the `nshots` can be recovered from the `samples` and I could actually make it a property. The `backend` I am not really sure how relevant it is, because if you load a new `MeasurementResult` object (which I find unlikely in any case) you are probably better off selecting the backend that you are currently using rather than the one that was in use when the object was dumped. Regarding the `circuit`, on the other hand, I would say it is safe to remove, and I plan doing it if everyone agrees. Now, the problem with the `measurement_gate` is that the `M` gate and the `MeasurementResult` are nested, they both contain an instance of the other one. Therefore, at some point you have to drop something. In this case, I decided to simplify the `MeasurementResult` serialization in favor of a more thorough `M` serialization. Mainly since In my experience with qibo, I had the impression that the `MeasurementResult` object is rarely used directly, and more often you access it through the measurement.

To be honest, as I see it now, the `MeasurementResult` is basically just a wrapper to the samples, and I would even argue that its dependence on the measurement gate is not necessary. In the end, I believe that the `measurement_gate` attribute  is only used to find the qubits (`measurement_gate.target_qubits`), at that point we would be better off building it just starting from the qubit indices.

Reply:
> Well, the `nshots` can be recovered from the `samples` and I could actually make it a property.

Great, let's do that

> The `backend` I am not really sure how relevant it is, because if you load a new `MeasurementResult` object (which I find unlikely in any case) you are probably better off selecting the backend that you are currently using rather than the one that was in use when the object was dumped.

Then, I would argue that the same could be done even immediately, dropping the `backend` attribute as well.

> Now, the problem with the `measurement_gate` is that the `M` gate and the `MeasurementResult` are nested, they both contain an instance of the other one.

That's worse than nested (which would be fine): they are cyclic.

> To be honest, as I see it now, the `MeasurementResult` is basically just a wrapper to the samples, and I would even argue that its dependence on the measurement gate is not necessary. In the end, I believe that the `measurement_gate` attribute is only used to find the qubits (`measurement_gate.target_qubits`), at that point we would be better off building it just starting from the qubit indices.

I would argue that the `MeasurementResult` methods could be turned into a library of functions. If it's convenient to group them together in a class, let's make this class containing a single attribute: the `samples`. Everything else can be received by argument to its methods, if needed.

In any case, this is a separate improvement. I'd suggest making it an issue.